prov/util/src/util_cq.c

```diff
@@ -301,11 +301,28 @@ ssize_t ofi_cq_sread(struct fid_cq *cq_fid, void *buf, size_t count,
 		const void *cond, int timeout)
 {
 	struct util_cq *cq;
+	uint64_t start;
+	int ret;
 
 	cq = container_of(cq_fid, struct util_cq, cq_fid);
 	assert(cq->wait && cq->internal_wait);
-	fi_wait(&cq->wait->wait_fid, timeout);
-	return ofi_cq_read(cq_fid, buf, count);
+	start = (timeout >= 0) ? fi_gettime_ms() : 0;
+
+	do {
+		if (timeout >= 0) {
+			timeout -= (int) (fi_gettime_ms() - start);
+			if (timeout <= 0)
+				return -FI_ETIMEDOUT;
```

Comment:
We need to adjust the timeout after calling ofi_cq_read().  With this change, if the user passes in a timeout of 0 (or even a small value), we'll abort without ever trying to read the CQ.

Reply:
Should we also call ofi_cq_read when fi_wait timeout expires?

Reply:
That likely depends on the return value of fi_wait() given a timeout of 0.  For a non-zero timeout where fi_wait expires, I don't think we need to read the CQ, as it should still be empty.  But it seems harmless to do so, so if the code structure is cleaner calling it, I don't see a problem.

Reply:
Maybe something like this would work:
```
do {
    ret = fi_cq_read();
    if (ret != -FI_EAGAIN)
        break;

    if (timeout >= 0) {
        ...
    }

    ret = fi_wait();
} while (ret == 0);  /* not sure about this check */
```

Reply:
Okay. I'll rearrange it based on above suggestion. Let's skip calling ofi_cq_read again if fi_wait timeout expires to keep it simple.